### PR TITLE
Clean up CalcPlrItemVals()

### PR DESCRIPTION
**Pull request by kphoenix137**
Massive refactor

### CODE REVIEW

File: Source/items.cpp [hunk 1]
```diff
@@ -2571,13 +2571,32 @@ void CalcPlrItemVals(Player &player, bool loadgfx)
 	int imana = 0; // increased mana
 
 	int spllvladd = 0; // increased spell level
-	int enac = 0;      // enhanced accuracy
+	int enac = 0;      // enhanced armor class
```
Comment (qndel) [inline Source/items.cpp:2574]: wrong

File: Source/items.cpp [hunk 2]
```diff
@@ -2571,13 +2571,32 @@ void CalcPlrItemVals(Player &player, bool loadgfx)
...
 
 	int fmin = 0; // minimum fire damage
 	int fmax = 0; // maximum fire damage
 	int lmin = 0; // minimum lightning damage
 	int lmax = 0; // maximum lightning damage
 
+	HeroClass pc = player._pClass;
+
+	ItemType leftHandItem = player.InvBody[INVLOC_HAND_LEFT]._itype;
+	ItemType rightHandItem = player.InvBody[INVLOC_HAND_RIGHT]._itype;
```
Comment (qndel) [inline Source/items.cpp:2584]: not rightHandItem but rightHandItemType, same for left

File: Source/items.cpp (hunk 1 above)
Comment (qndel) [reply Source/items.cpp:2574]: image
you even changed it in your other PR xD

File: Source/items.cpp [hunk 3]
```diff
@@ -2677,59 +2696,57 @@ void CalcPlrItemVals(Player &player, bool loadgfx)
...
 		}
 		player._pIAC += player._pLevel / 4;
-	} else {
+		break;
+	default:
 		player._pDamageMod = player._pLevel * player._pStrength / 100;
+		break;
 	}
-
+
```
Comment (qndel) [inline Source/items.cpp:2741]: pointless whitespace chance

Comment (qndel) [inline Source/items.cpp:2741]: pointless whitespace change

File: test/writehero_test.cpp [hunk 4]
```diff
@@ -294,8 +294,8 @@ static void AssertPlayer(Player &player)
 	ASSERT_EQ(CountBool(player._pSLvlVisited, NUMLEVELS), 0);
 	ASSERT_EQ(player._pNFrames, 20);
 	ASSERT_EQ(player._pWFrames, 8);
-	ASSERT_EQ(player._pAFrames, 0);
-	ASSERT_EQ(player._pAFNum, 0);
+	ASSERT_EQ(player._pAFrames, 18);
+	ASSERT_EQ(player._pAFNum, 10);
```
Comment (qndel) [inline test/writehero_test.cpp:298]: why

Comment (kphoenix137) [reply test/writehero_test.cpp:298]: Because test fails without it. Player is assigned frames in town now

Comment (qndel) [reply test/writehero_test.cpp:298]: I don't see any change related to it in this PR

Comment (kphoenix137) [reply test/writehero_test.cpp:298]: mistake then

File: Source/items.cpp [hunk 5]
```diff
@@ -2677,59 +2696,57 @@ void CalcPlrItemVals(Player &player, bool loadgfx)
...
+		if (IsAnyOf(ItemType::Staff, leftHandItemType, rightHandItemType) || (leftHandEmpty && rightHandEmpty)) {
 			player._pDamageMod = player._pLevel * (player._pStrength + player._pDexterity) / 150;
+		} else {
+			player._pDamageMod = player._pLevel * (player._pStrength + player._pDexterity) / 300;
 		}
-	} else if (player._pClass == HeroClass::Bard) {
-		if (player.InvBody[INVLOC_HAND_LEFT]._itype == ItemType::Sword || player.InvBody[INVLOC_HAND_RIGHT]._itype == ItemType::Sword)
+		break;
+	case HeroClass::Bard:
+		if (IsAnyOf(ItemType::Sword, leftHandItemType, rightHandItemType)) {
```
Comment (ephphatha) [inline Source/items.cpp:2711]: There are a lot of checks along the lines of "is the player holding a specific weapon in either hand", it'd be clearer to provide a function like `player.IsHoldingItem(ItemType)` where the definition looks like:

``` cpp
constexpr bool Player::IsHoldingItem(ItemType type) {
    return IsAnyOf(type, leftHandItem.type, rightHandItem.type);
}
```

Comment (ephphatha) [reply Source/items.cpp:2711]: Also consider whether this function should check that the items are actually usable. Some of the uses do, some don't, is this inconsistency desirable?

File: Source/items.cpp [hunk 6]
```diff
@@ -2578,6 +2578,25 @@ void CalcPlrItemVals(Player &player, bool loadgfx)
 	int lmin = 0; // minimum lightning damage
 	int lmax = 0; // maximum lightning damage
 
+	HeroClass pc = player._pClass;
+
+	ItemType leftHandItem = player.InvBody[INVLOC_HAND_LEFT]._itype;
```
[the remark this answers is not in the document]
Comment (ephphatha) [reply Source/items.cpp]: You addressed this for left/right hand items but it still applies to the chestItem variable

**Comment by kphoenix137** [comment]
Not sure why tests are failing

**Comment by ephphatha** [comment]
> Not sure why tests are failing

Part of that is my fault, I didn't explain the quirks with member function definitions.

linux_x86_64 and amigaos-m68k builds use GCC6. This version of GCC has a bug/overly restrictive implementation of constexpr checks which ends up rejecting attempts to compile a function as constexpr if the enclosing class is not a trivial class.

Player.CreatePlayer is failing because the test case is wrong. A rogue starts with a bow with 1-4 damage, max damage should have been 4 (but the logic was failing in CreatePlayer because item._istatflag is unintialised😂 ). This might be the same reason the writehero test is failing, I'm not sure why it's ending up with a lower than expected hp value.

**Comment by kphoenix137** [comment]
> > Not sure why tests are failing
> 
> Part of that is my fault, I didn't explain the quirks with member function definitions.
> 
> linux_x86_64 and amigaos-m68k builds use GCC6. This version of GCC has a bug/overly restrictive implementation of constexpr checks which ends up rejecting attempts to compile a function as constexpr if the enclosing class is not a trivial class.
> 
> Player.CreatePlayer is failing because the test case is wrong. A rogue starts with a bow with 1-4 damage, max damage should have been 4 (but the logic was failing in CreatePlayer because item._istatflag is unintialised😂 ). This might be the same reason the writehero test is failing, I'm not sure why it's ending up with a lower than expected hp value.

How do I fix? Remove constexpr?

**Comment by AJenbo** [comment]
Looks like this needs a little rebase

File: Source/items.cpp [hunk 7]
```diff
@@ -2901,7 +2852,114 @@ void CalcPlrItemVals(Player &player, bool loadgfx)
...
+	CalcPlrResistances(player, flags, fireRes, lightRes, magicRes);
+	CalcPlrLifeMana(player, vitality, magic, life, mana);
+	player._pIFMinDam = minFireDam;
+	player._pIFMaxDam = maxFireDam;
+	player._pILMinDam = minLightDam;
+	player._pILMaxDam = maxLightDam;
+
+	const bool holdsShield = CalcPlrBlockFlag(player);
+
+	CalcPlrGraphics(player, GetPlrAnimWeaponId(player, holdsShield), GetPlrAnimArmorId(player), loadgfx);
```
Comment (AJenbo) [inline Source/items.cpp:2961]: Feels odd to have `CalcPlrBlockFlag()` return true if the player is holding a shield. None of the other calc function return values and if anything it should return the block flag.

Let move add a check for holding shield inside `GetPlrAnimWeaponId()` instead.
```suggestion
	CalcPlrBlockFlag(player);
	CalcPlrGraphics(player, GetPlrAnimWeaponId(player), GetPlrAnimArmorId(player), loadgfx);
```

File: Source/items.cpp [hunk 8]
```diff
@@ -2559,318 +2559,266 @@ void InitItems()
 	initItemGetRecords();
 }
 
-void CalcPlrItemVals(Player &player, bool loadgfx)
+int GetBonusAC(auto &item)
```
Comment (AJenbo) [inline Source/items.cpp:2562]:
```suggestion
int GetBonusAC(ItemPack &item)
```

File: Source/items.cpp [hunk 9]
```diff
@@ -2559,318 +2559,266 @@ void InitItems()
...
 	}
 
-	const uint8_t playerLevel = player.getCharacterLevel();
+	return 0;
+}
 
-	if (mind == 0 && maxd == 0) {
-		mind = 1;
-		maxd = 1;
+void CalcPlrDamage(auto &player, int minDamage, int maxDamage)
```
Comment (AJenbo) [inline Source/items.cpp:2576]:
```suggestion
void CalcPlrDamage(Player &player, int minDamage, int maxDamage)
```

File: Source/items.cpp [hunk 10]
```diff
@@ -2559,318 +2559,266 @@ void InitItems()
...
+			minDamage = std::max(minDamage, playerLevel / 2);
+			maxDamage = std::max<int>(maxDamage, playerLevel);
 		}
 	}
 
+	player._pIMinDam = minDamage;
+	player._pIMaxDam = maxDamage;
+}
+
+void CalcPlrPrimaryStats(auto &player, int strength, int &magic, int dexterity, int &vitality)
```
Comment (AJenbo) [inline Source/items.cpp:2598]:
```suggestion
void CalcPlrPrimaryStats(Player &player, int strength, int &magic, int dexterity, int &vitality)
```

File: Source/items.cpp [hunk 11]
```diff
@@ -2559,318 +2559,266 @@ void InitItems()
...
+		vitality -= 2 * playerLevel;
+	}
+
+	player._pStrength = std::max(0, strength + player._pBaseStr);
+	player._pMagic = std::max(0, magic + player._pBaseMag);
+	player._pDexterity = std::max(0, dexterity + player._pBaseDex);
+	player._pVitality = std::max(0, vitality + player._pBaseVit);
+}
 
+void CalcPlrLightRadius(auto &player, int lrad)
```
Comment (AJenbo) [inline Source/items.cpp:2619]:
```suggestion
void CalcPlrLightRadius(Player &player, int lrad)

```

File: Source/items.cpp [hunk 12]
```diff
@@ -2559,318 +2559,266 @@ void InitItems()
...
-	} else if (player._pClass == HeroClass::Bard) {
-		if (player.InvBody[INVLOC_HAND_LEFT]._itype == ItemType::Sword || player.InvBody[INVLOC_HAND_RIGHT]._itype == ItemType::Sword)
-			player._pDamageMod = playerLevel * (player._pStrength + player._pDexterity) / 150;
-		else if (player.InvBody[INVLOC_HAND_LEFT]._itype == ItemType::Bow || player.InvBody[INVLOC_HAND_RIGHT]._itype == ItemType::Bow) {
-			player._pDamageMod = playerLevel * (player._pStrength + player._pDexterity) / 250;
+void CalcPlrDamageMod(auto &player)
+{
+	const uint8_t playerLevel = player.getCharacterLevel();
+	const auto &leftHandItem = player.InvBody[INVLOC_HAND_LEFT];
+	const auto &rightHandItem = player.InvBody[INVLOC_HAND_RIGHT];
```
Comment (AJenbo) [inline Source/items.cpp:2634]:
```suggestion
void CalcPlrDamageMod(Player &player)
{
	const uint8_t playerLevel = player.getCharacterLevel();
	const ItemPack &leftHandItem = player.InvBody[INVLOC_HAND_LEFT];
	const ItemPack &rightHandItem = player.InvBody[INVLOC_HAND_RIGHT];
```

File: Source/items.cpp [hunk 13]
```diff
@@ -2559,318 +2559,266 @@ void InitItems()
...
 	}
+}
 
-	player._pISpells = spl;
-
-	EnsureValidReadiedSpell(player);
-
-	player._pISplLvlAdd = spllvladd;
-	player._pIEnAc = enac;
+void CalcPlrResistances(auto &player, ItemSpecialEffect iflgs, int fire, int lightning, int magic)
```
Comment (AJenbo) [inline Source/items.cpp:2682]:
```suggestion
void CalcPlrResistances(Player &player, ItemSpecialEffect iflgs, int fire, int lightning, int magic)
```

File: Source/items.cpp [hunk 14]
```diff
@@ -2559,318 +2559,266 @@ void InitItems()
...
 
-	player._pMagResist = std::clamp(mr, 0, MaxResistance);
-	player._pFireResist = std::clamp(fr, 0, MaxResistance);
-	player._pLghtResist = std::clamp(lr, 0, MaxResistance);
+	player._pMagResist = std::clamp(magic, 0, MaxResistance);
+	player._pFireResist = std::clamp(fire, 0, MaxResistance);
+	player._pLghtResist = std::clamp(lightning, 0, MaxResistance);
+}
 
+void CalcPlrLifeMana(auto &player, int vitality, int magic, int life, int mana)
```
Comment (AJenbo) [inline Source/items.cpp:2710]:
```suggestion
void CalcPlrLifeMana(Player &player, int vitality, int magic, int life, int mana)
```

File: Source/items.cpp [hunk 15]
```diff
@@ -2559,318 +2559,266 @@ void InitItems()
...
+}
 
-	player._pIFMinDam = fmin;
-	player._pIFMaxDam = fmax;
-	player._pILMinDam = lmin;
-	player._pILMaxDam = lmax;
+bool CalcPlrBlockFlag(auto &player)
+{
+	const auto &leftHandItem = player.InvBody[INVLOC_HAND_LEFT];
+	const auto &rightHandItem = player.InvBody[INVLOC_HAND_RIGHT];
```
Comment (AJenbo) [inline Source/items.cpp:2733]:
```suggestion
bool CalcPlrBlockFlag(Player &player)
{
	const ItemPack &leftHandItem = player.InvBody[INVLOC_HAND_LEFT];
	const ItemPack &rightHandItem = player.InvBody[INVLOC_HAND_RIGHT];
```

File: Source/items.cpp [hunk 16]
```diff
@@ -2559,318 +2559,266 @@ void InitItems()
...
-	if (player.InvBody[INVLOC_CHEST]._itype == ItemType::HeavyArmor && player.InvBody[INVLOC_CHEST]._iStatFlag) {
-		if (player._pClass == HeroClass::Monk && player.InvBody[INVLOC_CHEST]._iMagical == ITEM_QUALITY_UNIQUE)
-			player._pIAC += playerLevel / 2;
-		animArmorId = PlayerArmorGraphic::Heavy;
-	} else if (player.InvBody[INVLOC_CHEST]._itype == ItemType::MediumArmor && player.InvBody[INVLOC_CHEST]._iStatFlag) {
-		if (player._pClass == HeroClass::Monk) {
-			if (player.InvBody[INVLOC_CHEST]._iMagical == ITEM_QUALITY_UNIQUE)
+PlayerArmorGraphic GetPlrAnimArmorId(auto &player)
+{
+	auto &chestItem = player.InvBody[INVLOC_CHEST];
```
Comment (AJenbo) [inline Source/items.cpp:2790]:
```suggestion
PlayerArmorGraphic GetPlrAnimArmorId(Player &player)
{
	ItemPack &chestItem = player.InvBody[INVLOC_CHEST];
```

File: Source/items.cpp [hunk 17]
```diff
@@ -2559,318 +2559,266 @@ void InitItems()
...
 		}
-		animArmorId = PlayerArmorGraphic::Medium;
-	} else if (player._pClass == HeroClass::Monk) {
-		player._pIAC += playerLevel * 2;
 	}
 
+	return PlayerArmorGraphic::Light;
+}
+
+void CalcPlrGraphics(auto &player, PlayerWeaponGraphic animWeaponId, PlayerArmorGraphic animArmorId, bool loadgfx)
```
Comment (AJenbo) [inline Source/items.cpp:2820]:
```suggestion
void CalcPlrGraphics(Player &player, PlayerWeaponGraphic animWeaponId, PlayerArmorGraphic animArmorId, bool loadgfx)
```

File: Source/items.cpp [hunk 18]
```diff
@@ -2889,7 +2837,10 @@ void CalcPlrItemVals(Player &player, bool loadgfx)
 	} else {
 		player._pgfxnum = gfxNum;
 	}
+}
 
+void CalcPlrAuricBonus(auto &player)
```
Comment (AJenbo) [inline Source/items.cpp:2842]:
```suggestion
void CalcPlrAuricBonus(Player &player)

```

File: Source/items.cpp [hunk 19]
```diff
@@ -2901,7 +2852,114 @@ void CalcPlrItemVals(Player &player, bool loadgfx)
...
+
+	int8_t splLvlAdd = 0;
+	int targetAc = 0;
+
+	int minFireDam = 0;
+	int maxFireDam = 0;
+	int minLightDam = 0;
+	int maxLightDam = 0;
+
+	for (auto &item : player.InvBody) {
```
Comment (AJenbo) [inline Source/items.cpp:2897]:
```suggestion
	for (ItemPack &item : player.InvBody) {
```

File: Source/player.h [hunk 20]
```diff
@@ -894,6 +894,15 @@ struct Player {
 
 	/** @brief Checks if the player level is owned by local client. */
 	bool isLevelOwnedByLocalClient() const;
+
+	/** @brief Checks if the player is holding an item of the provided type, and is usable. */
+	bool isHoldingItem(ItemType type)
+	{
+		auto &leftHandItem = InvBody[INVLOC_HAND_LEFT];
+		auto &rightHandItem = InvBody[INVLOC_HAND_RIGHT];
```
Comment (AJenbo) [inline Source/player.h:902]:
```suggestion
		ItemPack &leftHandItem = InvBody[INVLOC_HAND_LEFT];
		ItemPack &rightHandItem = InvBody[INVLOC_HAND_RIGHT];
```

File: Source/items.cpp [hunk 21]
```diff
@@ -2559,318 +2559,266 @@ void InitItems()
...
-	if (player.InvBody[INVLOC_HAND_LEFT]._itype == ItemType::Shield && player.InvBody[INVLOC_HAND_LEFT]._iStatFlag) {
+	if (player.isHoldingItem(ItemType::Shield)) {
 		player._pBlockFlag = true;
 		holdsShield = true;
 	}
-	if (player.InvBody[INVLOC_HAND_RIGHT]._itype == ItemType::Shield && player.InvBody[INVLOC_HAND_RIGHT]._iStatFlag) {
-		player._pBlockFlag = true;
-		holdsShield = true;
+
+	return holdsShield;
```
Comment (AJenbo) [inline Source/items.cpp:2753]:
```cpp
	player._pBlockFlag = player._pBlockFlag || player.isHoldingItem(ItemType::Shield);
```

File: Source/items.cpp (hunk 7 above)
Comment (AJenbo) [inline Source/items.cpp:2961]:
```suggestion
	CalcPlrBlockFlag(player);

	CalcPlrGraphics(player, GetPlrAnimWeaponId(player), GetPlrAnimArmorId(player), loadgfx);

```

File: Source/player.h (hunk 20 above)
Comment (AJenbo) [inline Source/player.h:902]:
```suggestion
	bool isHoldingItem(const ItemType type)
	{
		const ItemPack &leftHandItem = InvBody[INVLOC_HAND_LEFT];
		const ItemPack &rightHandItem = InvBody[INVLOC_HAND_RIGHT];
```

**Review by AJenbo — CHANGES_REQUESTED** [review]
We prefer to only use `auto` wen the type is immediately clear via something like a static cast in the assignment, if it's something overly complex that doesn't really affect things such as a function reference or pair.

Other then that I noticed a few places that would be good to objects as const.

**Review by AJenbo — DISMISSED** [review]

**Review by AJenbo — APPROVED** [review]

**Review by AJenbo — APPROVED** [review]

**Pull request merged**